src/gpio.rs

```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
+                        ahb.rstr().modify(|_, w| w.$iopxrst().clear_bit());
+
+                        Parts {
+                            afrh: AFRH(()),
+                            afrl: AFRL(()),
+                            moder: MODER(()),
+                            ospeedr: OSPEEDR(()),
+                            otyper: OTYPER(()),
+                            pupdr: PUPDR(()),
+                            $(
+                                $pxi: $PXi {
+                                    gpio: $Gpiox,
+                                    index: $Ui::new(),
+                                    _mode: PhantomData,
+                                },
+                            )+
                         }
                     }
+                }
 
-                    /// Partially erased pin
-                    pub struct $PXx<MODE> {
-                        i: u8,
-                        _mode: PhantomData<MODE>,
-                    }
+                /// Opaque AFRH register
+                pub struct AFRH(());
 
-                    impl<MODE> $PXx<MODE> {
-                        /// Erases the port letter from the type
-                        ///
-                        /// This is useful when you want to collect the pins into an array where you
-                        /// need all the elements to have the same type
-                        pub fn downgrade(self) -> PXx<MODE> {
-                            PXx {
-                                i: self.i,
-                                gpio: Gpio::$GPIOX,
-                                _mode: self._mode,
-                            }
-                        }
+                impl Afr for AFRH {
+                    #[inline]
+                    fn afx(&mut self, i: u8, x: u8) {
+                        unsafe { modify_at!((*$GPIOX::ptr()).afrh, 4, i - 8, x as u32); }
                     }
+                }
 
-                    impl<MODE> OutputPin for $PXx<Output<MODE>> {
-                        type Error = Infallible;
-
-                        fn set_high(&mut self) -> Result<(), Self::Error> {
-                            // NOTE(unsafe, write) atomic write to a stateless register
-                            unsafe { (*$GPIOX::ptr()).bsrr.write(|w| w.bits(1 << self.i)) }
-                            Ok(())
-                        }
+                /// Opaque AFRL register
+                pub struct AFRL(());
 
-                        fn set_low(&mut self) -> Result<(), Self::Error> {
-                            // NOTE(unsafe, write) atomic write to a stateless register
-                            unsafe { (*$GPIOX::ptr()).bsrr.write(|w| w.bits(1 << (16 + self.i))) }
-                            Ok(())
-                        }
+                impl Afr for AFRL {
+                    #[inline]
+                    fn afx(&mut self, i: u8, x: u8) {
+                        unsafe { modify_at!((*$GPIOX::ptr()).afrl, 4, i, x as u32); }
```

Comment:
Consider let expression that labels what 4 is.

Reply:
The `modify_at` macro could need some explanation, but 4 is essentially the `bitwidth` of the field in the 32 bit register, which is changed by this invocation.